mesonpy/_tags.py

```diff
@@ -1,130 +1,124 @@
-        return None
-
-    @property
-    def abi(self) -> str:
-        return f'abi{self.abi_number}'
-
-    def __eq__(self, other: Any) -> bool:
-        return isinstance(other, self.__class__) and other.abi_number == self.abi_number
-
-    def __hash__(self) -> int:
-        return hash(str(self))
-
-
-class InterpreterTag(Tag):
-    def __init__(self, value: str) -> None:
-        parts = value.split('-')
-        if len(parts) < 2:
-            raise ValueError(
-                'Invalid PEP 3149 interpreter tag, expected at '
-                f'least 2 parts but got {len(parts)}'
-            )
-
-        # On Windows, file extensions look like `.cp311-win_amd64.pyd`, so the
-        # implementation part (`cpython-`) is different from Linux. Handle that here:
-        if parts[0].startswith('cp3'):
-            parts.insert(0, 'cpython')
-            parts[1] = parts[1][2:]  # strip 'cp'
-
-        self._implementation = parts[0]
-        self._interpreter_version = parts[1]
-        self._additional_information = parts[2:]
-
-        if self.implementation != 'cpython' and not self.implementation.startswith('pypy'):
-            raise NotImplementedError(f'Unknown Python implementation: {self.implementation}.')
-
-    @property
-    def implementation(self) -> str:
-        return self._implementation
-
-    @property
-    def interpreter_version(self) -> str:
-        return self._interpreter_version
-
-    @property
-    def additional_information(self) -> Sequence[str]:
-        return tuple(self._additional_information)
+import os
+import platform
+import sys
+import sysconfig
+
+from typing import Optional, Union
+
+
+# https://peps.python.org/pep-0425/#python-tag
+INTERPRETERS = {
+    'python': 'py',
+    'cpython': 'cp',
+    'pypy': 'pp',
+    'ironpython': 'ip',
+    'jython': 'jy',
+}
+
+
+def get_interpreter_tag() -> str:
+    name = sys.implementation.name
+    name = INTERPRETERS.get(name, name)
+    version = sys.version_info
+    return f'{name}{version[0]}{version[1]}'
+
+
+def _get_config_var(name: str, default: Union[str, int, None] = None) -> Union[str, int, None]:
+    value = sysconfig.get_config_var(name)
+    if value is None:
+        return default
+    return value
+
+
+def _get_cpython_abi() -> str:
+    version = sys.version_info
+    debug = pymalloc = ''
+    if _get_config_var('Py_DEBUG', hasattr(sys, 'gettotalrefcount')):
+        debug = 'd'
+    if sys.version_info < (3, 8) and _get_config_var('WITH_PYMALLOC', True):
+        pymalloc = 'm'
+    return f'cp{version[0]}{version[1]}{debug}{pymalloc}'
+
+
+def get_abi_tag() -> str:
+    abi = sysconfig.get_config_var('SOABI')
+    if not abi:
+        # CPython on Windows does not have a SOABI sysconfig variable.
+        assert sys.implementation.name == 'cpython'
+        return _get_cpython_abi()
+    if abi.startswith('cpython'):
+        return 'cp' + abi.split('-')[1]
+    if abi.startswith('pypy'):
+        return '_'.join(abi.split('-')[:2])
```

Comment:
No special handling is needed for PyPy AFACT. Am I missing something?

Reply:
I agree. However, the special handling is in `packaging` and I decided to keep it. I'm happy to remove it if we are positive that it is not needed.

Reply:
Where? The correspondent code AFAIK is

https://github.com/pypa/packaging/blob/606c71acce93d04e778cdbdea16231b36d6b870f/packaging/tags.py#L496-L499

Which only has special handling for CPython, as it's the only needed. For other implementations, the value falls back to the normalized version of `SOABI`.

Reply:
In fact, we won't generate the same tags as `packaging` if PyPy adds another section to `SOABI`, as this code is only taking the first two.

Reply:
We should not be using SOABI. It is unreliable. It's better to use EXT_SUFFIX, and parse from there. Packaging is doing this in https://github.com/pypa/packaging/pull/607. Because packaging's usage of this has been wrong for so long, other implementiaonts might have broken `SOABI`'s too, just to make packing work. While `EXT_SUFFIX` is much more likely to be correct, as it's used in other places. Both variables are documented in [PEP 3149](https://peps.python.org/pep-3149/), and are supposed to be identical, minus the leading dot and extension. Also, Python itself recommends `EXT_SUFFIX` if you try to use `SO`.

Reply:
Sorry, I meant `wheel` https://github.com/pypa/wheel/blob/814c2efe8e40051039c5a6de6945e04ecdd162ee/src/wheel/bdist_wheel.py#L96-L99 I've been staring at all these implementations for too long. I agree that the special handling is weird. I'll remove it.

Reply:
@henryiii I think I don't understand. I cannot parse what you are trying to say with this sentence

> Because packaging's usage of this has been wrong for so long, other implementations might have broken `SOABI`'s too, just to make packing work (and therefor pip, etc).

Can you please try to rephrase it?

If PEP3149 documents both variables to carry the same information, if they would get out of sync, that would be a bug of the Python implementation. Wouldn't it? Can we find a Python implementation that has the wrong `SOABI` but the right `EXT_SUFFIX`?

Reply:
> Can we find a Python implementation that has the wrong `SOABI` but the right `EXT_SUFFIX`?

Absolutely: PyPy. PyPy reports the wrong SOABI, it's missing the platform information. If you use it, you make an extension that can't be loaded (since EXT_SUFFIX is used for loading). When Packaging was adapted to use SOABI, it special cases CPython, so the "general" code was based on PyPy's mistake, so it assumes SOABI is wrong. So other implementations might report wrong SOABI's because it makes packing work. It's being fixed in packaging right now (to use EXT_SUFFIX and assume it is right), but until the current Pip is old enough to not matter, PyPy is stuck reporting the wrong SOABI.

If you report the wrong `EXT_SUFFIX`, setuptools and probably other things break. If you report the _right_ `SOABI` but you are not CPython, you might not be loadable by Pip currently (is being fixed in that linked PR).

Reply:
According to the PEP, `assert sysconfig.get_config_var('EXT_SUFFIX').split(".", 1)[0].lstrip(".") == sysconfig.get_config_var('SOABI')`, but it doesn't for PyPy, so using the former is best. For just the ABI, yes, that would suffice.

What's wrong with CPython 3.7?

```console
$ docker run --rm -it quay.io/pypa/manylinux2014_x86_64
[root@f00ca105d347 /]# python3.7
Python 3.7.14 (default, Sep 12 2022, 15:16:12)
[GCC 10.2.1 20210130 (Red Hat 10.2.1-11)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sysconfg
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'sysconfg'
>>> import sysconfig
>>> sysconfig.get_config_var("EXT_SUFFIX")
'.cpython-37m-x86_64-linux-gnu.so'
>>> sysconfig.get_config_var("SOABI")
'cpython-37m-x86_64-linux-gnu'
>>>
[root@f00ca105d347 /]# pypy3.7
Python 3.7.13 (7e0ae751533460d5f89f3ac48ce366d8642d1db5, Mar 29 2022, 06:03:31)
[PyPy 7.3.9 with GCC 10.2.1 20210130 (Red Hat 10.2.1-11)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>> import sysconfig
>>>> sysconfig.get_config_var("EXT_SUFFIX")
'.pypy37-pp73-x86_64-linux-gnu.so'
>>>> sysconfig.get_config_var("SOABI")
'pypy37-pp73'
```

Reply:
My recollection that CPython started adding more complex EXT_SUFFIX instead of just ".so", as part of the migration from 2.7 to the entire 3.x series, and that change, like most 3.x changes, happened "sometime during 3.0, 3.1, 3.2, or 3.3, although most people aren't sure which of them because it took several releases before the general populace considered it to be a viable replacement for 2.7".

Reply:
> Was it an early patch release of 3.7? Or something else?

CPython 3.7 on Windows just reports the extension. See #190.

Reply:
PEP 3149 was accepted in 2010, so that's what I would have figured too, which is why I was asking about why a fallback was needed for 3.7.

Reply:
See https://docs.python.org/3.8/whatsnew/changelog.html#python-3-8-7-final and https://github.com/python/cpython/issues/84006

Reply:
Ah, you said `.so`. Okay, sounds like a fallback is needed for Windows and Python 3.7, then.

Reply:
Yes, my bad. Sorry.

Reply:
Nope. Mayne I misunderstood the proposal, but looking at `EXT_SUFFIX` and assuming is correct does not fly.

CPython
```
>>> sysconfig.get_config_var('EXT_SUFFIX')
'.cpython-310-darwin.so'
>>> next(packaging.tags.sys_tags()).abi
'cp310'
```

PyPy:
```
>>>> next(packaging.tags.sys_tags()).abi
'pypy37_pp73'
>>>> sysconfig.get_config_var('EXT_SUFFIX')
'.pypy37-pp73-darwin.so'
```

Pyston:
```
>>> print(sysconfig.get_config_var("EXT_SUFFIX"))
'.pyston-23-x86_64-linux-gnu.so'
>>> next(packaging.tags.sys_tags()).abi
'pyston_23_x86_64_linux_gnu'
```

Also, this is the beautiful code currently proposed for `packaging` in 
```python
def _generic_abi() -> List[str]:
    """Return the ABI tag based on EXT_SUFFIX"""
    # ext_suffix is something like this. We want to keep the parts which are
    # related to the ABI and remove the parts which are related to the
    # platform:
    # - linux:   '.cpython-310-x86_64-linux-gnu.so' => cp310
    # - mac:     '.cpython-310-darwin.so'           => cp310
    # - win:     '.cp310-win_amd64.pyd'             => cp310
    # - pypy:    '.pypy38-pp73-x86_64-linux-gnu.so' => pypy38-pp73
    # - graalpy: '.graalpy-38-native-x86_64-darwin.dylib'
    #                                               => graalpy-38-native

    ext_suffix = _get_config_var("EXT_SUFFIX", warn=True)
    assert ext_suffix[0] == "."
    _, soabi, ext = ext_suffix.split(".")
    if soabi.startswith("cpython"):
        abi = "cp" + soabi.split("-")[1]
    elif soabi.startswith("pypy"):
        abi = "-".join(soabi.split("-")[:2])
    elif soabi.startswith("graalpy"):
        abi = "-".join(soabi.split("-")[:3])
    elif soabi:
        abi = soabi
    else:
        return []
    return [_normalize_string(abi)]
```
which is many things but not generic. Lovely.

Reply:
I'm going to stick to `SOABI` just because it takes lees code. Is really a `sysconfig.get_interpreter_abi()` function too much to ask?

Reply:
What is the result of SOABI in those cases? It should be `cpython-310-darwin`, `pypy37-pp73-darwin` (but actually is `pypy37-pp73`), and `pyston-23-x86_64-linux-gnu`, correct?

Reply:
By "assuming it is correct", I mean that it's the correct extension suffix. `pypy37-pp73.so` for example is not a valid suffix, pypy can't load it. I didn't mean the wheel abi name is "better" from `EXT_SUFFIX`, I just meant `SOABI` is sometimes wrong, while `EXT_SUFFIX` is correct. You still need to transform it for wheel tag names in either case - and it's exactly the same transform unless SOABI is buggy. And if SOABI is buggy, the implementation will hopefully try to change it (as PyPy will in a year or two).

Reply:
> I don't understand how going from special casing PyPy to special casing PyPy and GrallVM Python while still getting it wrong for Pyston is an improvement

The pyston developers are welcome to add to that code or provide a comment as to what needs to be added. In a perfect world, sysconfig would provide `get_python_tag()`, `get_abi_tag()` and `get_platform_tag()`, but it doesn't, so here we are. The goal of the PR you mentioned is to fix a possible issue when PyPy fixes its SOABI value at some time in the future. It was not intended to make the code generic (which is impossible as long as packaging needs to be backward-compatible) or to cover all implementations (that is up to other projects to contribute where needed).

Both pip and wheel now use `packaging/tags.py`. It is a shame this functionality is replicated in other places.

Reply:
> The pyston developers are welcome to add to that code or provide a comment as to what needs to be added.

How `packaging` generates tags for Pyston cannot be changed for compatibility: compatibility of existing wheel releases with systems upgraded to an eventual future `packaging` and compatibility of systems with an old packaging with wheels generated on a system with an eventual future `packaging`.

> Both pip and wheel now use `packaging/tags.py`. It is a shame this functionality is replicated in other places.

Should `packaging` start to provide `get_python_tag()`, `get_abi_tag()` and `get_platform_tag()` functions, we may consider it. Till then the API is inconvenient to use. The algorithm is set in stone for all existing Python implementations (see above) and new Python implementations do not yet pup up like mushrooms, so keeping it in sync is not much work at all.

Reply:
> What is the result of `SOABI` in those cases? It should be `cpython-310-darwin`, `pypy37-pp73-darwin` (but actually is `pypy37-pp73`), and `pyston-23-x86_64-linux-gnu`, correct?

Correct.